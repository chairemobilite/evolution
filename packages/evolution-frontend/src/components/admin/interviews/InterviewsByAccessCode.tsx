/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { useTranslation } from 'react-i18next';
import Loadable from 'react-loadable';
import Loader from 'react-spinners/HashLoader';

import { InterviewContext } from '../../../contexts/InterviewContext';
import InputString from 'chaire-lib-frontend/lib/components/input/InputString';
import { Link, useLocation, useParams } from 'react-router';
import { _booleish } from 'chaire-lib-common/lib/utils/LodashExtensions';

const loader = function Loading() {
    return <Loader size={30} color={'#aaaaaa'} loading={true} />;
};

const InterviewsComponent = Loadable({
    // TODO: move this to componentDidMount like in Monitoring.tsx
    loader: () => import('./InterviewSearchList'),
    loading: loader
});

const InterviewsByAccessCode: React.FC = () => {
    const location = useLocation();
    const params = useParams();
    const { t } = useTranslation('admin');

    const urlSearch = new URLSearchParams(location.search);
    const [currentCode, setCurrentCode] = React.useState(params.accessCode);
    const [createNewIfNoData] = React.useState(_booleish(urlSearch.get('autoCreate')) === true);
    const { state, dispatch } = React.useContext(InterviewContext);

    React.useEffect(() => {
        if (params.accessCode !== state.response.accessCode) {
            dispatch({ type: 'update_response', response: { accessCode: params.accessCode } });
        }
    }, [params.accessCode, state.response]);

    return (
        <div className="admin">
            <div className="survey-section__content apptr__form-container">
                <InputString
                    id="accessCodeSearchInput"
                    value={currentCode}
                    onValueUpdated={(newValue) => setCurrentCode(newValue.value)}
                ></InputString>
                <Link to={`/interviews/byCode/${currentCode}`}>
                    <button type="button" className={'survey-section__button button blue small'}>
                        {t('admin:interviewSearch:SearchByCode')}
                    </button>
                </Link>
            </div>
            <InterviewsComponent autoCreateIfNoData={createNewIfNoData} queryData={urlSearch} />
        </div>
    );
};

export default InterviewsByAccessCode;
