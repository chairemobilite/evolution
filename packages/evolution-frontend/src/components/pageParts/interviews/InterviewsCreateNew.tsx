/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { createBrowserHistory } from 'history';
import Loader from 'react-spinners/HashLoader';
import { InterviewContext } from '../../../contexts/InterviewContext';

const InterviewsCreateNew: React.FunctionComponent = () => {
    const [inProgress, setInProgress] = React.useState(false);
    const { state, dispatch } = React.useContext(InterviewContext);

    // Function to fetch data from the server, with paging and filtering
    const createUser = React.useCallback(async () => {
        if (inProgress || state.status !== 'creating') {
            return;
        }

        // Set in progress state
        setInProgress(true);

        // Make a query string from the filters
        try {
            const response = await fetch('/api/interviews/createNew', {
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json'
                },
                credentials: 'include',
                method: 'POST',
                body: JSON.stringify({
                    createUser: state.username,
                    responses: state.responses
                })
            });
            const history = createBrowserHistory({ forceRefresh: true });
            if (response.status === 200) {
                const jsonData = await response.json();
                if (jsonData.interviewUuid) {
                    history.push(`/survey/edit/${jsonData.interviewUuid}/`);
                    dispatch({ type: 'success', interviewUuid: jsonData.interviewUuid });
                    return;
                }
                console.log('Server did not return a proper interview id', jsonData);
                // Goto proper error page
            }
            // TODO Go to proper status code page
            console.log('Bad response code from server', response.status);
        } catch (error) {
            // TODO Go to proper error page
            console.error(`Error fetching user data from server: ${error}`);
        }

        // Cancel in progress state, only if the page was not redirected
        setInProgress(false);
    }, []);

    React.useEffect(() => {
        createUser();
    }, []);

    return (
        <div className="admin">
            {inProgress && <Loader size={30} color={'#aaaaaa'} loading={true} />}
            {!inProgress && 'error'}
        </div>
    );
};

/** Component used to create new interviews by users other than the main user */
export default InterviewsCreateNew;
