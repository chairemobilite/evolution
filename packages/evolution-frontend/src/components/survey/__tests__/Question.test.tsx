/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _cloneDeep from 'lodash.clonedeep';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import each from 'jest-each';
import { mount } from 'enzyme';
import { render } from '@testing-library/react'
import { axe, toHaveNoViolations } from 'jest-axe';
expect.extend(toHaveNoViolations)

import { interviewAttributes } from '../../inputs/__tests__/interviewData.test';
import Question from '../Question';
import { WidgetStatus } from '../../../services/interviews/interview';

// Mock the react-datepicker files to avoid jest compilation errors
jest.mock('react-datepicker/dist/react-datepicker.css', () => {});
// Mock the react-input-range files to avoid jest compilation errors
jest.mock('react-input-range/src/js/input-range/default-class-names', () => ({
    activeTrack: 'input-range__track input-range__track--active',
    disabledInputRange: 'input-range input-range--disabled',
    inputRange: 'input-range',
    labelContainer: 'input-range__label-container',
    maxLabel: 'input-range__label input-range__label--max',
    minLabel: 'input-range__label input-range__label--min',
    slider: 'input-range__slider',
    sliderContainer: 'input-range__slider-container',
    track: 'input-range__track input-range__track--background',
    valueLabel: 'input-range__label input-range__label--value',
}));

jest.mock('react-input-range/lib/css/index.css', () => {});

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

const commonWidgetConfig = {
    type: 'question' as const,
    twoColumns: true,
    path: 'home.region',
    containsHtml: true,
    label: {
        fr: `Texte en fran√ßais`,
        en: `English text`
    }
};

const defaultWidgetStatus: WidgetStatus<unknown, unknown, unknown, unknown> = {
    path: commonWidgetConfig.path,
    isVisible: true,
    modalIsOpen: false,
    isDisabled: false,
    isCollapsed: false,
    isEmpty: false,
    isCustomEmpty: false,
    isValid: true,
    isResponded: true,
    isCustomResponded: false,
    errorMessage: undefined,
    currentUpdateKey: 1,
    value: null
};

each([
    ['InputSelect', { ...commonWidgetConfig, inputType: 'select', choices: [{ label: 'choice 1', value: 'c1' }, { label: 'choice 2', value: 'c2' }]}, 'c2'],
    ['InputRadio', { ...commonWidgetConfig, inputType: 'radio', choices: [{ label: 'choice 1', value: 'c1' }, { label: 'choice 2', value: 'c2' }]}, 'c2'],
    ['InputCheckbox', { ...commonWidgetConfig, inputType: 'checkbox', choices: [{ label: 'choice 1', value: 'c1' }, { label: 'choice 2', value: 'c2' }]}, ['c2']],
    ['InputMultiselect', { ...commonWidgetConfig, inputType: 'multiselect', choices: [{ label: 'choice 1', value: 'c1' }, { label: 'choice 2', value: 'c2' }]}, 'c2'],
    ['InputButton', { ...commonWidgetConfig, inputType: 'button', choices: [{ label: 'choice 1', value: 'c1' }, { label: 'choice 2', value: 'c2' }]}, 'c2'],
    ['InputDatePicker', { ...commonWidgetConfig, inputType: 'datePicker', minDate: new Date('2023-05-01'), maxDate: new Date('2023-06-01') }, new Date('2023-05-24 10:00:00 GMT-0400')],
    ['InputRange', { ...commonWidgetConfig, inputType: 'slider', minValue: 3, maxValue: 10 }],
    ['InputText', { ...commonWidgetConfig, inputType: 'text' }, 'foo'], // This test needs a value
    ['InputString', { ...commonWidgetConfig, inputType: 'string' }],
    ['InputTime', { ...commonWidgetConfig, inputType: 'time', minTimeSecondsSinceMidnight: 3600, maxTimeSecondsSinceMidnight: 7200, minuteStep: 10 }, 3660], // This test needs a value
]).describe('Question with widget %s', (_widget, widgetConfig, value: unknown = undefined) => {

    const widgetStatus = _cloneDeep(defaultWidgetStatus);
    widgetStatus.value = value as any;

    test('Render widget', () => {
        
        const wrapper = TestRenderer.create(
            <Question
                path='home.region'
                section='test'
                loadingState={0}
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={widgetStatus}
                startUpdateInterview={() => { /* Nothing to do */}}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Widget accessibility', async () => {
        const { container} = render(
            <Question
                path='home.region'
                section='test'
                loadingState={0}
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={widgetStatus}
                startUpdateInterview={() => { /* Nothing to do */}}
            />
        );
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});

describe('With help popup and link', () => {
    const helpContent = 'Help content';
    const widgetConfig = {
        ...commonWidgetConfig,
        inputType: 'string',
        helpPopup: {
            title: 'Help title',
            content: jest.fn().mockReturnValue(helpContent)
        }
    };

    test('Test widget with help', () => {
        const widgetStatus = _cloneDeep(defaultWidgetStatus);
        widgetStatus.value = 'test';
        
        const wrapper = TestRenderer.create(
            <Question
                path='home.region'
                section='test'
                loadingState={0}
                widgetConfig={widgetConfig as any}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={widgetStatus}
                startUpdateInterview={() => { /* Nothing to do */}}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Open modal', () => {
        const widgetStatus = _cloneDeep(defaultWidgetStatus);
        widgetStatus.value = 'test';
    
        const questionWidget = mount(<Question
            path='home.region'
            section='test'
            loadingState={0}
            widgetConfig={widgetConfig as any}
            interview={interviewAttributes}
            user={userAttributes}
            widgetStatus={widgetStatus}
            startUpdateInterview={() => { /* Nothing to do */}}
        />);

        // Find and click on the Geocode button, to return multiple values
        expect(widgetConfig.helpPopup.content).not.toHaveBeenCalled();
        const helpButton = questionWidget.find('.helper-popup');
        const modal = questionWidget.find('.react-modal');
        expect(modal.children().length).toEqual(0);
        expect(helpButton.getDOMNode<HTMLButtonElement>().textContent).toEqual(widgetConfig.helpPopup.title);
        helpButton.simulate('click');
    
        // The modal should be opened now
        questionWidget.update();
        expect(widgetConfig.helpPopup.content).toHaveBeenCalledTimes(1);
        const modalAfterClick = questionWidget.find('.react-modal');
        expect(modalAfterClick).toMatchSnapshot();
        
    });
});

describe('With error message', () => {
    const widgetConfig = {
        ...commonWidgetConfig,
        inputType: 'string',
    };

    test('Error message on visible widget', () => {
        const widgetStatus = _cloneDeep(defaultWidgetStatus);
        widgetStatus.errorMessage = 'error test';
        widgetStatus.isValid = false;

        const wrapper = TestRenderer.create(
            <Question
                path='home.region'
                section='test'
                loadingState={0}
                widgetConfig={widgetConfig as any}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={widgetStatus}
                startUpdateInterview={() => { /* Nothing to do */}}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

});

// TODO: Test with modal and other Question props