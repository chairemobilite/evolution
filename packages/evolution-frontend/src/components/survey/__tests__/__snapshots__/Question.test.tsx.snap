// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal widget Widget is not visible 1`] = `<div />`;

exports[`Modal widget Widget is visible 1`] = `
<div>
  <div
    class="ReactModal__Overlay react-modal-overlay"
  >
    <div
      aria-label=""
      aria-modal="true"
      class="ReactModal__Content react-modal"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="apptr__form-container two-columns question-type-string question-filled question-valid"
        style="position: relative;"
      >
        <div
          class="apptr__form-label-container two-columns"
        >
          <label
            for="survey-question__home.region"
          >
            <div>
              English text
            </div>
          </label>
        </div>
        <div
          class="apptr__form-input-container two-columns"
        >
          <input
            autocomplete="none"
            class="apptr__form-input apptr__input-string input-large"
            id="survey-question__home.region"
            maxlength="255"
            name="survey-question__home.region"
            style="text-transform: none;"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question with widget InputButton Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-button question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <div
        class="survey-question__input-button-group-container"
      >
        <div
          class="survey-question__input-button-input-container"
        >
          <button
            class="survey-section__button button green large"
            id="survey-question__home.region__input-button__c1"
            type="button"
          >
            choice 1
          </button>
        </div>
        <div
          class="survey-question__input-button-input-container"
        >
          <button
            class="survey-section__button button green large"
            id="survey-question__home.region__input-button__c2"
            type="button"
          >
            choice 2
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question with widget InputCheckbox Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-checkbox question-filled question-valid"
    style="position: relative;"
  >
    <fieldset>
      <div
        class="apptr__form-label-container two-columns"
      >
        <legend>
          <div
            class="apptr__form__label-standalone"
          >
            <div>
              English text
            </div>
          </div>
        </legend>
      </div>
      <div
        class="apptr__form-input-container two-columns"
      >
        <div
          class="survey-question__input-checkbox-group-container"
        >
          <div
            class="survey-question__input-checkbox-input-container"
          >
            <div
              class="label-input-container"
            >
              <input
                class="input-checkbox"
                id="survey-question__home.region_c1__input-checkbox__c1"
                name="survey-question__home.region_c1"
                type="checkbox"
                value="c1"
              />
              <label
                for="survey-question__home.region_c1__input-checkbox__c1"
              >
                <span>
                  choice 1
                </span>
              </label>
            </div>
          </div>
          <div
            class="survey-question__input-checkbox-input-container checked"
          >
            <div
              class="label-input-container"
            >
              <input
                checked=""
                class="input-checkbox"
                id="survey-question__home.region_c2__input-checkbox__c2"
                name="survey-question__home.region_c2"
                type="checkbox"
                value="c2"
              />
              <label
                for="survey-question__home.region_c2__input-checkbox__c2"
              >
                <span>
                  choice 2
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`Question with widget InputDatePicker Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-datePicker question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <div
        class="survey-question__input-date-picker-container"
      >
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              autocomplete="none"
              class="apptr__form-input input-date-picker input-large"
              id="survey-question__home.region"
              name="survey-question__home.region"
              placeholder="ClickToSelectDate"
              type="text"
              value="24/05/2023"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question with widget InputMultiselect Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-multiselect question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <div
        class="apptr__form__label-standalone"
        id="survey-question__home.region_label"
      >
        <div>
          English text
        </div>
      </div>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <div
        class="survey-question__input-multiselect-container"
      >
        <div
          class="react-select-container css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="react-select__control css-13cymwt-control"
          >
            <div
              class="react-select__value-container react-select__value-container--is-multi react-select__value-container--has-value css-3w2yfm-ValueContainer"
            >
              <div
                class="react-select__multi-value css-1p3m7a8-multiValue"
              >
                <div
                  class="react-select__multi-value__label css-wsp0cs-MultiValueGeneric"
                >
                  <div
                    style="display: flex; align-items: center;"
                  >
                    <div
                      style="border-radius: 6px; height: 6px; width: 6px; margin-right: 6px;"
                    />
                    choice 2
                  </div>
                </div>
                <div
                  aria-label="Remove [object Object]"
                  class="react-select__multi-value__remove css-12a83d4-MultiValueRemove"
                  role="button"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="14"
                    viewBox="0 0 20 20"
                    width="14"
                  >
                    <path
                      d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="react-select__input-container css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-labelledby="survey-question__home.region_label"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="react-select__input"
                  id="react-select-2-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class="react-select__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div>
            <input
              name="survey-question__input-multiselect-home.region"
              type="hidden"
              value="c2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question with widget InputRadio Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-radio question-filled question-valid"
    style="position: relative;"
  >
    <fieldset>
      <div
        class="apptr__form-label-container two-columns"
      >
        <legend>
          <div
            class="apptr__form__label-standalone"
          >
            <div>
              English text
            </div>
          </div>
        </legend>
      </div>
      <div
        class="apptr__form-input-container two-columns"
      >
        <div
          class="survey-question__input-radio-group-container"
        >
          <div
            class="survey-question__input-radio-input-container"
          >
            <div
              class="label-input-container"
            >
              <input
                class="input-radio"
                id="survey-question__home.region_c1__input-radio__c1"
                name="survey-question__home.region_c1"
                type="radio"
                value="c1"
              />
              <label
                for="survey-question__home.region_c1__input-radio__c1"
              >
                <span>
                  choice 1
                </span>
              </label>
            </div>
          </div>
          <div
            class="survey-question__input-radio-input-container checked"
          >
            <div
              class="label-input-container"
            >
              <input
                checked=""
                class="input-radio"
                id="survey-question__home.region_c2__input-radio__c2"
                name="survey-question__home.region_c2"
                type="radio"
                value="c2"
              />
              <label
                for="survey-question__home.region_c2__input-radio__c2"
              >
                <span>
                  choice 2
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`Question with widget InputRadioNumber Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-radioNumber question-filled question-valid"
    style="position: relative;"
  >
    <fieldset>
      <div
        class="apptr__form-label-container two-columns"
      >
        <legend>
          <div
            class="apptr__form__label-standalone"
          >
            <div>
              English text
            </div>
          </div>
        </legend>
      </div>
      <div
        class="apptr__form-input-container two-columns"
      >
        <div
          class="survey-question__input-radio-group-container"
        >
          <div
            class="survey-question__input-radio-input-container"
          >
            <div
              class="label-input-container"
            >
              <input
                class="input-radio"
                id="survey-question__home.region_1__input-radio__1"
                name="survey-question__home.region_1"
                type="radio"
                value="1"
              />
              <label
                for="survey-question__home.region_1__input-radio__1"
              >
                <span>
                  1
                </span>
              </label>
            </div>
          </div>
          <div
            class="survey-question__input-radio-input-container"
          >
            <div
              class="label-input-container"
            >
              <input
                class="input-radio"
                id="survey-question__home.region_2__input-radio__2"
                name="survey-question__home.region_2"
                type="radio"
                value="2"
              />
              <label
                for="survey-question__home.region_2__input-radio__2"
              >
                <span>
                  2
                </span>
              </label>
            </div>
          </div>
          <div
            class="survey-question__input-radio-input-container"
          >
            <div
              class="label-input-container"
            >
              <input
                class="input-radio"
                id="survey-question__home.region_3__input-radio__3"
                name="survey-question__home.region_3"
                type="radio"
                value="3"
              />
              <label
                for="survey-question__home.region_3__input-radio__3"
              >
                <span>
                  3
                </span>
              </label>
            </div>
          </div>
          <div
            class="survey-question__input-radio-input-container checked"
          >
            <div
              class="label-input-container"
            >
              <input
                checked=""
                class="input-radio"
                id="survey-question__home.region_4__input-radio__4"
                name="survey-question__home.region_4"
                type="radio"
                value="4"
              />
              <label
                for="survey-question__home.region_4__input-radio__4"
              >
                <span>
                  4+
                </span>
              </label>
            </div>
          </div>
          <div
            style="display: flex; align-items: center; gap: 1rem;"
          >
            <label
              for="survey-question__home.regionover-max"
            >
              <span>
                SpecifyAboveLimit:
              </span>
            </label>
            <input
              class="apptr__form-input apptr__input-string input-large"
              id="survey-question__home.regionover-max"
              min="4"
              name="survey-question__home.regionover-max"
              pattern="[0-9]*"
              style="width: 6rem;"
              type="number"
              value="4"
            />
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`Question with widget InputRange Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-slider question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <div
        class="apptr__form__label-standalone"
        id="survey-question__home.region_label"
      >
        <div>
          English text
        </div>
      </div>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <div
        class="survey-question__input-range_input_container_with_labels input-blank"
      >
        <div
          aria-disabled="false"
          class="input-range"
        >
          <span
            class="input-range__label input-range__label--min"
          >
            <span
              class="input-range__label-container"
            >
              3
            </span>
          </span>
          <div
            class="input-range__track input-range__track--background"
          >
            <div
              class="input-range__track input-range__track--active"
              style="left: 0%; width: 0%;"
            />
            <span
              class="input-range__slider-container"
              style="position: absolute; left: 0%;"
            >
              <span
                class="input-range__label input-range__label--value"
              >
                <span
                  class="input-range__label-container"
                >
                  3
                </span>
              </span>
              <div
                aria-labelledby="survey-question__home.region_label"
                aria-valuemax="10"
                aria-valuemin="3"
                aria-valuenow="3"
                class="input-range__slider"
                draggable="false"
                role="slider"
                tabindex="0"
              />
            </span>
          </div>
          <span
            class="input-range__label input-range__label--max"
          >
            <span
              class="input-range__label-container"
            >
              10
            </span>
          </span>
          <input
            name="survey-question__home.region"
            type="hidden"
            value="3"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question with widget InputSelect Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-select question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <div
        class="survey-question__input-select-container"
      >
        <select
          id="survey-question__home.region"
          style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis; max-width: 100%;"
        >
          <option
            class="input-select-option"
            value=""
          />
          <option
            class="input-select-option"
            value="c1"
          >
            choice 1
          </option>
          <option
            class="input-select-option"
            value="c2"
          >
            choice 2
          </option>
          <optgroup
            label=""
          />
        </select>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question with widget InputString Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-string question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <input
        autocomplete="none"
        class="apptr__form-input apptr__input-string input-large"
        id="survey-question__home.region"
        maxlength="255"
        name="survey-question__home.region"
        style="text-transform: none;"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Question with widget InputText Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-text question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <textarea
        autocomplete="off"
        class="apptr__form-input input-text"
        id="survey-question__home.region"
        maxlength="2000"
        rows="5"
      >
        foo
      </textarea>
    </div>
  </div>
</div>
`;

exports[`Question with widget InputTime Render widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-time question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <div
        class="survey-question__input-select-container survey-question__input-time-container"
      >
        <select
          id="survey-question__home.region"
        >
          <option
            class="input-select-option"
            value=""
          />
          <option
            value="3600"
          >
            1:00 am
          </option>
          <option
            value="4200"
          >
            1:10 am
          </option>
          <option
            value="4800"
          >
            1:20 am
          </option>
          <option
            value="5400"
          >
            1:30 am
          </option>
          <option
            value="6000"
          >
            1:40 am
          </option>
          <option
            value="6600"
          >
            1:50 am
          </option>
          <option
            value="7200"
          >
            2:00 am
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;

exports[`With error message Error message on visible widget 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-string question-filled question-invalid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
      <p
        class="apptr__form-error-message"
      >
        <span>
          error test
        </span>
      </p>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <input
        autocomplete="none"
        class="apptr__form-input apptr__input-string input-large"
        id="survey-question__home.region"
        maxlength="255"
        name="survey-question__home.region"
        style="text-transform: none;"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`With help popup and link Open modal 1`] = `
<div
  aria-label="Help title"
  aria-modal="true"
  class="ReactModal__Content react-modal"
  role="dialog"
  tabindex="-1"
>
  <div>
    <h3
      class="center"
    >
      <markdown
        remarkplugins="remark-gfm,[object Object]"
      >
        Help title
      </markdown>
    </h3>
    <markdown
      class="help-popup"
      remarkplugins="remark-gfm,[object Object]"
    >
      Help content
    </markdown>
    <div
      class="center"
    >
      <button
        class="button blue"
      >
        OK
      </button>
    </div>
  </div>
</div>
`;

exports[`With help popup and link Test widget with help 1`] = `
<div>
  <div
    class="apptr__form-container two-columns question-type-string question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
      <div>
        <button
          class="button helper-popup blue small"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-circle-question faIconLeft"
            data-icon="circle-question"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM169.8 165.3c7.9-22.3 29.1-37.3 52.8-37.3l58.3 0c34.9 0 63.1 28.3 63.1 63.1c0 22.6-12.1 43.5-31.7 54.8L280 264.4c-.2 13-10.9 23.6-24 23.6c-13.3 0-24-10.7-24-24l0-13.5c0-8.6 4.6-16.5 12.1-20.8l44.3-25.4c4.7-2.7 7.6-7.7 7.6-13.1c0-8.4-6.8-15.1-15.1-15.1l-58.3 0c-3.4 0-6.4 2.1-7.5 5.3l-.4 1.2c-4.4 12.5-18.2 19-30.6 14.6s-19-18.2-14.6-30.6l.4-1.2zM224 352a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"
              fill="currentColor"
            />
          </svg>
          Help title
        </button>
      </div>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <input
        autocomplete="none"
        class="apptr__form-input apptr__input-string input-large"
        id="survey-question__home.region"
        maxlength="255"
        name="survey-question__home.region"
        style="text-transform: none;"
        type="text"
        value="test"
      />
    </div>
  </div>
</div>
`;

exports[`With joining questions Joining next 1`] = `
<div>
  <div
    class="apptr__form-container apptr__form-join-next two-columns question-type-string question-filled question-valid"
    style="position: relative;"
  >
    <div
      class="apptr__form-label-container two-columns"
    >
      <label
        for="survey-question__home.region"
      >
        <div>
          English text
        </div>
      </label>
    </div>
    <div
      class="apptr__form-input-container two-columns"
    >
      <input
        autocomplete="none"
        class="apptr__form-input apptr__input-string input-large"
        id="survey-question__home.region"
        maxlength="255"
        name="survey-question__home.region"
        style="text-transform: none;"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
